Transitioning project to Gradle.

floppypanda opened the pull request:
This pull request contains the changes necessary to build Elephant using Gradle. Elephant's current build system is driven solely through Eclipse, which is fine, but I believe Gradle can provide this project with many additional benefits. Some of these benefits include:

- Building and running Elephant from the command line on any of the major operating systems.
- Integration with Eclipse, IntelliJ IDEA, NetBeans and any Java IDE that has a Gradle plugin.
- Easy and efficient dependency management in which:
    - JAR files are no longer committed to the repository and are instead downloaded automatically, along with their dependencies, from the repositories we configure in `build.gradle`.
    - Upgrades simply consist of changing the version number of a dependency in `build.gradle` and rebuilding.
- Gradle tasks pave the way for running automated tests, generating documentation, and code linting.
- Gradle facilitates the use of free continuous integration tools such as Travis CI.
- README.md and  CONTRIBUTING.md can be updated with command line steps for quickly getting up and running with Elephant development (more than happy to contribute here if needed).
- And much more...

This may appear to be a large change based on the number of files changed, but can be summarized in just a few bullet points:
- Moved `src/com` to `src/main/java/com` so that source files could be found by the Gradle build.
- Moved `src/images`, `src/notes`, and `src/style` to `src/main/resources/images`, `src/main/resources/notes` and `src/main/resources/style` respectively. Gradle includes files placed in `src/main/resources` as resources of the Elephant JAR.
- Added files and directories generated by Gradle (`gradle`, `gradlew`, `gradlew.bat`) as well as those that allow us to configure Gradle (`build.gradle` `settings.gradle`).
- Updated Eclipse files (`.classpath`, `.profile`) to be representative of how they look following importing Elephant into Eclipse as a Gradle project.
- Removing `build.xml` and `exportJar.xml` as the purpose of these is now fulfilled by `build.gradle`.

I'm more than willing to discuss changes or additions to this pull request and feedback is very much welcome. Some things you should know when providing feedback include:
- These changes were tested on a Linux system, so it's definitely worth determining if they function the same on macOS and Windows. Fortunately, these are only build system changes, with no code being touched, so they can be verified by executing `./gradlew build` and then `./gradlew run` on the command line (use `./gradlew.bat` for Windows).
- If you're using Eclipse, import Elephant as a Gradle project and then debug or run the application to verify changes.
- I'm curious about your thoughts on integrating `build_mac_app.xml`, `build_mac_app_bundle_jre.xml` and `elephantexe.xml` with Gradle. Should I convert them to Gradle tasks and include them in `build.gradle`? I understand what they are doing, but `elephantexe.xml` appears to be based off your local file system and I'm not sure if the others are out of date.
- You may notice that there are fewer dependencies in `build.gradle` than there were JAR files in the `libs` directory. This is due to Gradle's automatic dependency resolution.
- In the line `compile 'io.github.dheid:fontchooser:2.3'` within `build.gradle`, I had to use version 2.3 of fontchooser instead of 2.4 as 2.4 is not yet provided by the repositories available to Gradle.
- The last of the original JAR files that were committed to the Elephant repository, `lib/com.pinktwins.rtf.jar`, remains for the time being until I add a task for pulling down and building a JAR from the `com.pinktwins.rtf` directory automatically. That would be included in a later pull request, if this one is accepted.

I know this is a lot to digest, but I hope to hear back from you about these changes soon.

floppypanda committed:
Moving source files to src/main/java for Gradle compatibility.

floppypanda committed:
Renaming icon images directory to "icons".

floppypanda committed:
Moving images directory to base directory.

floppypanda committed:
Moving notes directory to base directory.

floppypanda committed:
Moving style directory to base directory.

floppypanda committed:
Moving images to resource directory.

floppypanda committed:
Moving notes to resource directory.

floppypanda committed:
Moving styles to resource directory.

floppypanda committed:
Adding Gradle files.

floppypanda committed:
Ignoring .gradle.

floppypanda committed:
Moving style files to appropriate directory.

floppypanda committed:
Removing build.xml and exportJar.xml in favor of Gradle.

floppypanda committed:
Renaming libs to lib.

floppypanda committed:
Removing libraries that are now automatically downloaded by Gradle.

floppypanda committed:
Updating location of com.pinktwins.rtf.jar.

floppypanda committed:
Alphabetizing plugins.

floppypanda committed:
Updating .classpath and .project as generated by Gradle.

jusu commented:
Thanks, I can see and appreciate the effort and thoughtfulness put into this. This change seems to be a step towards facilitating more contributions and automating builds, testing etc to avoid regressions or issues that more frequent contributions might bring. But I'm not sure, or I'm questioning, if my goals for this project are in line with that. Here are some thoughts:

- You can currently build from command line using Ant and build.xml. Is this not working for you, or are there showstoppers there I'm not aware of?

- My understanding is having those jar dependencies in the repo makes building and packaging Elephant a bit difficult for linux distributions, is this the main reason to move them to a central repo? I would prefer to use these exact versions of these libs for the next 20 years unless there's a specific reason to update, and I'd expect them to be unavailable at some point, as happened with fontchooser already. I see not depending on an external service as a positive for this project, and I don't even know what alfresco.com is... Well, just trying to outline why the libs are there in the first place. In short, they are where I want them to be? Does that make sense?

- How would you build meaningful tests for what is mostly a ui-heavy application?

- Can ant / build.xml and the gradle build coexist?

Thanks again. I should explain my approach for this project better in the readme/contributions files.

floppypanda commented:
The bullets below address your questions and concerns. Constructive feedback is provided and is not meant to be disparaging to yourself or the project. It's all in the interest of helping out. Feel free to let me know if you have any additional questions:

**Command Line Builds** 
- After checking again, it turns out that building from the command line is possible, but it requires two commands, `ant` and then `ant exportJar.xml` to produce an executable JAR file. I wasn't aware of this when familiarizing myself with the project as there were a number of XML build files in the root directory whose usage and order of application are not documented. I instead imported into Eclipse, being under the impression that this was the convention for the project or that certain build steps might be dependent on the IDE. 
    - In this light, I would argue that Gradle doesn't enable command line builds to occur where they couldn't previously, but that it makes them simpler for contributors to use. Building and running Gradle, regardless of the number of tasks, can be simplified to `./gradlew build` and `./gradlew run` whereas we currently have `ant`, `ant exportJar.xml`, and then `java -jar build/elephantapp.jar`.
    - It could be argued, however, that `build.xml` could be improved so that running `ant` builds and exports the Elephant JAR file and `ant run` runs `java -jar build/elephantapp.jar` along with any default JVM arguments we may specify.
    - The takeaway is that command line builds work currently, but could be improved even if we choose not to introduce Gradle.

**JAR Dependencies in Git**
- Including JAR files in version control does not make building and packaging difficult on Linux distributions and is not the main reason to move them to a central repository. Some reasons to move them to a central repository include:
    - Reduction of repository cloning time
        - Certainly the JARs will need to be downloaded sometime at or before build time, but newcomers might want to inspect the code locally before committing to expending bandwidth on them.
   - Trust
       - Imagine that a contributor adds a feature that requires a new JAR to be committed to `libs`. How do we know they didn't embed malicious code into the JAR? The classes and method signatures in the JAR may be the same as the official one, but there is no way to know unless a checksum is applied and compared to that of the official. Centralized repositories combined with Gradle help us ensure that we are getting the JARs that we mean to.
       - On the flip side, how can we convince contributors and users of Elephant to trust that we haven't included JARs in the project repository that include malware? Again, myself and others can verify the checksum of each JAR, but this is cumbersome. In the end, the dependencies list in `build.gradle` places the burden of trust on a 3rd party that is under constant scrutiny by the security community.
       - Understand that there is some debate regarding the security of centralized repositories [Maven Security](https://stackoverflow.com/questions/24967270/how-do-we-know-we-can-trust-the-maven-central-repository#24987915) (be sure to take a look at both sides), but that doesn't mean that our current position is inherently more secure. If you're concerned about that and want to keep code and dependencies separate, we could always make our own [Maven Repo on GitHub](https://stackoverflow.com/a/14013645).
    - Updating
        - I understand that you'd prefer to keep the dependency versions the same, but the project might not last for 20 years if you do so. Nearly all of the JAR files included in `libs` are severely out of date with a number of them including significant security vulnerabilities. If we want to increase adoption of Elephant and become a secure and stable alternative to Evernote, ongoing change must occur. In particular, the PDF Renderer is enough out of date that it no longer has a repository for its source code and [a fork](https://github.com/katjas/PDFrenderer) has been created to address its issues.
       - Allowing dependencies to reach the point where they are no longer available and, if open source, can no longer have their source code verified isn't a best practice for an open source software project. Ideally we would upgrade or transition to an alternative before that could happen. The Maven Central Repository keeps around old versions of JARs for decades so we shouldn't worry about them disappearing.

**Other**
- The decision of using `fontchooser2.3` instead of `fontchooser2.4` was made as the latter is so new that it hasn't yet been added to the Maven Central Repository, not because it will never be added. 
- The Alfresco repository was used as it was one of the only credible repositories that contains a copy of `pdfrenderer`. If you don't feel comfortable with using the Alfresco Maven Repository, we could include the `pdfrenderer` JAR in the repository and reference it like `com.pinktwins.rtf.jar` is from Gradle.
- Automated functional tests can be created for testing UI-heavy applications. Functional testing frameworks such as [TestFX](https://github.com/TestFX/TestFX) apply actual mouse clicks and keyboard strokes (in a similar way to [Robot](https://docs.oracle.com/javase/7/docs/api/java/awt/Robot.html)) to GUIs to ensure they perform as intended. It may even be possible to use the OS specific builds on Travis CI to run functional tests on each of the major operating systems automatically. Gradle would make adding this dependency and running the tests much easier.
- The Ant and Gradle builds can co-exist, but it doesn't make much sense to include both and have them do the same thing.

jusu commented:
Actually the old java.net/Swinglabs-based pdfrenderer is not used anymore. Elephant uses renderer from Apache PDFBox. I should remove the old jar and code.

jusu commented:
Taking this one step at a time, can you comment on this pull req:?
https://github.com/jusu/Elephant/pull/105

floppypanda commented:
> Actually the old java.net/Swinglabs-based pdfrenderer is not used anymore. Elephant uses renderer from Apache PDFBox. I should remove the old jar and code.

I'm glad to hear that `pdfrenderer` is no longer used, as it was one of the dependencies I was most concerned about. I ran a vulnerability scan on my fork of Elephant using https://snyk.io/ (a free security tool on the GitHub marketplace) and I encourage you to do the same in order to view the results. Unfortunately, Synk doesn't appear to have a shareable link feature for reports otherwise I would have provided you with direct access to it.

Note that I had to merge the Gradle changes into the `master` branch of my fork in order for Synk to properly analyze dependencies for vulnerabilities (it is unable to do so based on Ant files alone, or on a branch other than `master` apparently), so you should run Synk on Elephant after you've merged #105.

jusu committed:
Update build.gradle

Update guava + tika, add 2 jai-imageio jars.

With these changes, Snyk gives clean output:

```
Testing /Users/jusu/Desktop/Elephant-master...

Organisation:      jzzz
Package manager:   gradle
Target file:       build.gradle
Open source:       no
Project path:      /Users/jusu/Desktop/Elephant-master

✓ Tested 145 dependencies for known vulnerabilities, no vulnerable paths found.
```

jusu commented:
This commit https://github.com/jusu/Elephant/pull/104/commits/b67075035136ca28dfcbdebe166bcb0fe6052a6a
will produce clean 'snyk test' output.

Thanks for helping out! I'll be updating my versions as well.

floppypanda commented:
Thank you for taking the time to address the various aspects of this pull request! I've enjoyed collaborating with you on this and I'm glad that I've been able to help out. I'm closing this pull request as the changes have been applied elsewhere.
